lib/patchelf/alt_saver.rb

```diff
@@ -785,29 +965,46 @@ def phdrs_by_type(seg_type)
       end
     end
 
-    # To be used when the section header does not exist.
-    def dummy_shdr
-      ELFTools::Structs::ELF_Shdr.new(endian: endian, elf_class: elf_class)
+    # Returns a blank shdr if the section doesn't exist.
+    def find_or_create_section_header(rsec_name)
+      shdr = find_section(rsec_name)&.header
+      shdr ||= ELFTools::Structs::ELF_Shdr.new(endian: endian, elf_class: elf_class)
+      shdr
     end
 
-    def write_replaced_sections(cur_off, start_addr, start_offset)
-      sht_no_bits = ELFTools::Constants::SHT_NOBITS
-
+    def overwrite_replaced_sections
       # the original source says this has to be done separately to
       # prevent clobbering the previously written section contents.
       @replaced_sections.each do |rsec_name, _|
         shdr = find_section(rsec_name)&.header
         next unless shdr
 
-        with_buf_at(shdr.sh_offset) { |b| b.fill('X', shdr.sh_size) } if shdr.sh_type != sht_no_bits
+        next if shdr.sh_type == ELFTools::Constants::SHT_NOBITS
+
+        with_buf_at(shdr.sh_offset) { |b| b.fill('X', shdr.sh_size) }
       end
+    end
+
+    def write_section_aligment(shdr)
+      return if shdr.sh_type == ELFTools::Constants::SHT_NOTE && shdr.sh_addralign <= @section_alignment
+
+      shdr.sh_addralign = @section_alignment
+    end
+
+    def overlapping?(start1, end1, start2, end2)
+      (start1 >= start2 && start1 < end2) || (end1 > start2 && end1 <= end2)
```

Comment:
Hmm, what if this case
```
   |--------------------------------|
start1                             end1
                  |----------|
                start2      end2
```

Reply:
Good point. The logic matches upstream. The error I made here was generalising the function when in fact `start1` had to be sections and `start2` had to be segments. I've renamed it as such.

I guess the point of what happens if a SHT_NOTE overlaps a PT_NOTE still stands, but something must have gone wrong to reach that point, and it's not clear whether we are picking up the correct PT_NOTE in that case anyway with the sizes wildly mismatching, so just skipping over that PT_NOTE is probably remains safest.

Hard to tell without seeing a binary that behaves that way, which in theory shouldn't exist.